core/lib/zksync_core/src/reorg_detector/mod.rs

```diff
@@ -86,8 +93,7 @@ trait MainNodeClient: fmt::Debug + Send + Sync {
 
     async fn l2_block_hash(&self, number: L2BlockNumber) -> EnrichedClientResult<Option<H256>>;
 
-    async fn l1_batch_root_hash(&self, number: L1BatchNumber)
-        -> EnrichedClientResult<Option<H256>>;
+    async fn l1_batch_root_hash(&self, number: L1BatchNumber) -> EnrichedClientResult<StateHash>;
```

Comment:
why not Result<H256,SomeErr>?

Reply:
`EnrichedClientResult` is an alias for `Result<_, EnrichedClientError>` in the same spirit as `anyhow::Result<_>` is an alias of `Result<_, anyhow::Error>`. IMO, the `Result` suffix in the type name makes it easy to parse.

Reply:
I know that. I meant  EnrichedClientResult<Result<H256,SomeError>> or embed the MissingBatch, MissingHash in the EnrichedClientError.

Reply:
Sorry for misunderstanding. Subjectively, Non-`H256` values don't feel like an error from the client perspective (more like an extended `Option`), and changing `EnrichedClientError` for this specific purpose looks awkward because it's defined in an upstream crate and is intended for general-purpose errors.

Reply:
Ok, I feel the opposite:
* anything different than root hash is not a happy path, therefore an error is a reasonable result
* Result IS an extended Option, since Option<X> is isomorphic to Result<X,()>. And it is the intended rust interpretation, given the obvious correspondence between Option and Result APIs

Reply:
I also feel that it would be clearer to make l1_batch_root_hash cancel aware and move the retries to the implementation. This would simplify the control flow.